src/v/raft/mux_state_machine.h

```diff
@@ -126,54 +126,49 @@ class mux_state_machine : public state_machine {
 
 private:
     using promise_t = expiring_promise<std::error_code>;
-    // promises used to wait for result of state applies, keyed by offser
-    // returned in replication result (i.e. last batch end offset)
-    using container_t
-      = absl::node_hash_map<model::offset, expiring_promise<std::error_code>>;
 
     ss::future<> apply(model::record_batch b) final;
+    class replicate_units {
+    public:
+        explicit replicate_units(mux_state_machine<T...>& stm)
+          : _stm(stm) {
+            _stm._pending++;
+            _units = 1;
+        }
+
+        replicate_units(replicate_units&& other) noexcept
+          : _stm(other._stm)
+          , _units(std::exchange(other._units, 0)) {}
+
+        replicate_units(const replicate_units&) noexcept = delete;
+
+        replicate_units& operator=(replicate_units&& o) noexcept {
+            _stm = o._stm;
```

Comment:
Can you explain the semantics of assigning the state machine through this reference?

Should it be a pointer? Would that allow default move and assignment?

Reply:
i just want to make sure that when `replicate_units` is move assigned it will not subtract `_pending` counter. Indeed stm should not be copyable or movable